feat: add webhooks

boogsbunny opened the pull request:
## Background
This PR adds the ability to verify and construct [Stripe webhook events](https://docs.stripe.com/webhooks). This permits us to handle Stripe webhook events like customer’s bank confirms a payment, a customer disputes a charge, a recurring payment succeeds, or when collecting subscription payments. This goes hand in hand with the session endpoint to create user subscriptions for example.

We also switched from `yason` to `jzon` for our JSON library and thus handle object instantions for hash tables instead of plists now.

## Changes
- Add webhook event errors.
- Add webhook secret variable to store the secret.
- Add webhook event object with the types specified.
- Add [ironclad](https://github.com/sharplispers/ironclad) as dependency.
- Replace [yason](https://github.com/phmarek/yason) with [jzon](https://github.com/Zulu-Inuoe/jzon) for the JSON library.
- Add function to validate webhook event.
- Add function to construct webhook event object.
- Add test system.
- Add webhook test suite.
- Add balance test suite.
- Add balance transaction test suite.
- Add card test suite.
- Add charge test suite.
- Add common test suite.
- Add coupon test suite.
- Add credit note test suite.
- Add customer balance transaction test suite.
- Add invoice test suite.
- Add invoice item test suite.
- Add payout test suite.
- Add plan test suite.
- Add product test suite.
- Add refund test suite.
- Add session test suite.
- Add subscription test suite.
- Add subscription item test suite.
- Add token test suite.
- Update `decode-list` now `decode-hash-table` to handle hash table including vector of hash tables.
- Update endpoints that fetch a list of objects passing in vector as type instead of list.
- Update `decode-timestamp` to handle null symbol and string.
- Update `initialize-instance` on stripe objects to handle null symbol.
- Update all other `initialize-instance` methods to handle hash table instead of plist.

## Testing
I've tested this locally and added unit tests.

## Notes
- I've added a `:key` parameter called  `ignore-api-version-mismatch `to the `construct-webhook-event` function since our current API version is not up to date.
- I've built this using the [API documentation](https://docs.stripe.com/api/events) and cross-referencing the [golang library](https://github.com/stripe/stripe-go/blob/master/webhook/client.go).
- Here's another [lisp library](https://github.com/K1D77A/lisp-pay/blob/main/src/stripe/webhooks.lisp) that validates webhooks for reference.

## Considerations
- [x] I'm still wrapping my head around writing a test suite, though that can be added after. I'm thinking of using [fiveam](https://github.com/lispci/fiveam)?
- [ ] Should we add all [event types](https://github.com/stripe/stripe-go/blob/26539500cc8267f40f25c0c4ff7235fbdeec6ebf/event.go#L19) as constants in this PR?
- [x] Should we change the JSON library to [jzon](https://github.com/Zulu-Inuoe/jzon) in this PR? Now, we're passing in the payload and `yason:parse` expects a string or a stream. Many people use [clack](https://github.com/fukamachi/clack), which uses `vector-streams`. It's [straightforward](https://github.com/Zulu-Inuoe/jzon?tab=readme-ov-file#jzonparse) to pass that as the argument for the `payload` parameter to the `construct-webhook-event`, but not if we use `yason`.

aadcg commented:
@boogsbunny overall, it looks good to me. Maybe @jmercouris has something to add.

-----


> I'm still wrapping my head around writing a test suite, though that can be added after. I'm thinking of using [fiveam](https://github.com/lispci/fiveam)?

It's always better later then never. But some things are easier to do when they're still fresh.

Personally, I really like https://github.com/AccelerationNet/lisp-unit2, but you should use whatever you're most comfortable with.


-----



Regarding the JSON library, if we move forward with the change, then I'd say that a test suite becomes key. Otherwise there may be regressions.


----

Regarding adding types, follow your intuition.

boogsbunny commented:
So I've just added tests for webhooks!

I suppose the only other thing that can be added here are also adding tests to verify `jzon:parse` works as expected having replaced `yason:parse`. What do you think @aadcg?

aadcg commented:
@boogsbunny yes, sounds good. As to ensure that the migration to the new library doesn't break any of logic (unrelated to webhooks). Thanks!

boogsbunny commented:
I've written tests for the `charge` object and noticed that it fails when parsing the API response in the `define-query` macro if the type parameter is a list:

```
             ,@(case type
                 (list `((decode-list ,response)))
                 ((nil) `(,response))
                 (t `((make-instance ',type :data ,response))))))))))
```

The issue occurs because `response` is a hash table where the list of objects is now a vector of hash tables, and the `decode-list` function in `common.lisp` currently expects a list of objects but fails to create the objects from the vector of hash tables.

Right now, we succeed at creating an object but fail for endpoints where we receive a list of them.

My proposal is to modify the `decode-list` function to handle the vector of hash tables instead of a list. This would involve changing the `initialize-instance` methods in some files from using `destructuring-bind` to use `gethash`, allowing the code to handle the vector of hash tables correctly.

How does this direction sound?

I can push my code either way for you to get a better picture of the changes.

aadcg commented:
@boogsbunny sounds good!

boogsbunny commented:
The PR is ready for review @aadcg!

I've added tests to cover a good portion of the functions now, removed deprecated fields that no longer exist on objects, marked certain endpoints as deprecated since Stripe no longer supports them (we can remove those files), and updated the `initialize-method)` to handle the vector of hash tables now instead of lists.

Here's the output from running the tests locally:

```
<0:CL-USER> (stripe/tests:run)

Running test suite BALANCE-TESTS
 Running test RETRIEVE-BALANCE ...........
 Did 11 checks.
    Pass: 11 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite BALANCE-TRANSACTION-TESTS
 Running test RETRIEVE-BALANCE-TRANSACTION ...........
 Running test LIST-BALANCE-HISTORY ...
 Did 14 checks.
    Pass: 14 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite CARD-TESTS
 Running test CREATE-CARD ..........
 Running test LIST-CARDS ........
 Did 18 checks.
    Pass: 18 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite CHARGE-TESTS
 Running test LIST-CHARGES ............
 Did 12 checks.
    Pass: 12 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite COUPON-TESTS
 Running test CREATE-COUPON ........
 Running test LIST-COUPONS ......
 Did 14 checks.
    Pass: 14 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite CREDIT-NOTE-TESTS
 Running test CREATE-CREDIT-NOTE ...............
 Running test LIST-CREDIT-NOTES ................
 Did 31 checks.
    Pass: 31 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite CUSTOMER-BALANCE-TRANSACTION-TESTS
 Running test CREATE-CUSTOMER-BALANCE-TRANSACTION ...........
 Running test LIST-CUSTOMER-BALANCE-TRANSACTIONS ............
 Did 23 checks.
    Pass: 23 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite INVOICE-TESTS
 Running test CREATE-INVOICE ............................
 Running test LIST-INVOICES .............................
 Did 57 checks.
    Pass: 57 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite INVOICE-ITEM-TESTS
 Running test CREATE-INVOICE-ITEM .............
 Running test LIST-INVOICE-ITEMS ..............
 Did 27 checks.
    Pass: 27 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite PAYOUT-TESTS
 Running test CREATE-PAYOUT .............
 Running test LIST-PAYOUTS ..............
 Did 27 checks.
    Pass: 27 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite PLAN-TESTS
 Running test CREATE-PLAN ..........
 Running test LIST-PLANS ...........
 Did 21 checks.
    Pass: 21 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite PRODUCT-TESTS
 Running test CREATE-PRODUCT .............
 Running test LIST-PRODUCTS ..............
 Did 27 checks.
    Pass: 27 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite REFUND-TESTS
 Running test CREATE-REFUND ........
 Running test LIST-REFUNDS .........
 Did 17 checks.
    Pass: 17 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite SESSION-TESTS
 Running test RETRIEVE-SESSION ..
 Did 2 checks.
    Pass: 2 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite SUBSCRIPTION-TESTS
 Running test CREATE-SUBSCRIPTION ...................
 Running test LIST-SUBSCRIPTIONS ....................
 Did 39 checks.
    Pass: 39 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite SUBSCRIPTION-ITEM-TESTS
 Running test CREATE-SUBSCRIPTION-ITEM .....
 Running test LIST-SUBSCRIPTION-ITEMS ......
 Did 11 checks.
    Pass: 11 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite TOKEN-TESTS
 Running test CREATE-CARD-TOKEN ......
 Did 6 checks.
    Pass: 6 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)


Running test suite WEBHOOK-TESTS
 Running test VALIDATE-WEBHOOK-PAYLOAD-VALID .
 Running test VALIDATE-WEBHOOK-PAYLOAD-INVALID-SIGNATURE .
 Running test VALIDATE-WEBHOOK-PAYLOAD-TIMESTAMP-TOO-OLD .
 Running test CONSTRUCT-WEBHOOK-EVENT-VALID .......
 Running test CONSTRUCT-WEBHOOK-EVENT-INVALID-HEADER .
 Running test CONSTRUCT-WEBHOOK-EVENT-NO-VALID-SIGNATURE .
 Running test CONSTRUCT-WEBHOOK-EVENT-API-VERSION-MISMATCH .
 Running test PARSE-SIGNATURE-HEADER-VALID ...
 Running test PARSE-SIGNATURE-HEADER-INVALID .
 Did 17 checks.
    Pass: 17 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)

T
NIL
NIL
```

Fear not, the majority of the diff is just the tests haha :^)

aadcg reviewed (commented):
@boogsbunny thanks!

Please ensure that docstrings don't go beyond 80 chars per line. 

Let's keep the git history simple (squash needed): one commit for webhooks, another for the json library migration (including changes motivates by it), and another for the test suite.

I ran the tests locally and they pass. I'd suggest setting the dependencies of `stripe/tests` to `stripe` and `fiveam` alone. You're using `babel:string-to-octets` but it can be replaced by `dexador.util:ascii-string-to-octets` since dexador is an explicit dependency of `stripe`.

boogsbunny commented:
done, thanks!

aadcg commented on src/webhook.lisp:
```diff
@@ -0,0 +1,105 @@
+  "Default tolerance for signatures in seconds.")
+
+(defconstant +signing-version+ "v1"
+  "Current signature version.")
```
`+signing-version+` raises an error on `load-system`. It's a known issue when using `defconstant` so just fix it as per below. `+default-tolerance+` gets changed too, for consistency. 

```diff
-(defconstant +default-tolerance+ 300
-  "Default tolerance for signatures in seconds.")
+(alex:define-constant +default-tolerance+ 300
+  :documentation "Default tolerance for signatures in seconds.")
 
-(defconstant +signing-version+ "v1"
-  "Current signature version.")
+(alex:define-constant +signing-version+ "v1" :test #'string=
+  :documentation "Current signature version.")
```

aadcg commented on src/common.lisp:
```diff
@@ -6,22 +6,69 @@
 (defvar *api-key*)
 
+(defvar *webhook-secret*
+  "The endpoint's secret, used to generate [webhook signatures](https://docs.stripe.com/webhooks/signatures).")
```
Please keep the docstring within 100 chars, by dropping the link to the stripe documentation.

aadcg commented on src/common.lisp:
```diff
@@ -6,22 +6,69 @@
+  (cond
+    ((eq unix-time 'null) nil)
+    (unix-time (local-time:unix-to-timestamp unix-time))
+    (t nil)))
```
Take a closer look here. What happens when `unix-time` is of type, say, string?

aadcg commented on src/common.lisp:
```diff
@@ -6,22 +6,69 @@
+
+If unix-time is the symbol 'null, the function returns nil. This is because the
+jzon library returns the 'null symbol for null values in JSON, but unix-time
+cannot use that value.
```
Your goal here is to document that the function guards against a possible pitfall. But you don't need to mention a specific JSON library, i.e. you may frame it more generally.

aadcg commented on src/package.lisp:
```diff
@@ -1,13 +1,16 @@
+  (:use #:cl)
   (:local-nicknames
+   (#:alex #:alexandria)
+   (#:jzon #:com.inuoe.jzon)
```
Note that this change belong to the JSON migration commit, not the one that adds webhooks support.

aadcg commented on src/webhook.lisp:
```diff
@@ -0,0 +1,105 @@
+                       (error "Slot ~S does not exist in class ~S"
+                              (intern (concatenate 'string "%" (string key)) :keyword)
+                              (class-name (class-of instance))))))
+        (let ((created (slot-value instance '%created)))
```
If you move the `let` construct as the first branch of the `if`, then you don't need `progn` anymore.

aadcg commented on src/webhook.lisp:
```diff
@@ -0,0 +1,105 @@
+      (progn
+        (loop for key being the hash-keys of args
+              for value being the hash-values of args
+              do (let ((slot-name (find-symbol (concatenate 'string "%" (string key)) :stripe)))
```
Tip: when there's a `let` inside of a `do` clause, then it probably means that you're not leveraging the `for` clause.

Compare these:

```lisp
(loop for i in '(1 2)
      do (let ((arg (* i 2))) (print (+ i arg))))

(loop for i in '(1 2)
      for arg = (* i 2)
      do (print (+ i arg)))
```

Note that you can also use conditionals within `for` in a more idiomatic way.

aadcg commented on src/webhook.lisp:
```diff
@@ -0,0 +1,105 @@
+                                  ignore-tolerance
+                                  ignore-api-version-mismatch)
+  "Initialize an Event object from a JSON webhook payload, validating the Stripe-Signature header
+using the specified signing secret and options."
```
The first line of a docstring should fit within one line and be 100 chars at most.

Applies in others cases in this PR.

aadcg commented on src/webhook.lisp:
```diff
@@ -0,0 +1,105 @@
+
+(defun construct-webhook-event (payload header secret
+                                &key (tolerance +default-tolerance+)
+                                  ignore-tolerance
```
Wouldn't it be cleaner to have `tolerance` alone? `(tolerance +default-tolerance+ tolerance-provided-p)` and, in case `tolerance-provided-p` is T and `tolerance` is NIL, then ignore it.

aadcg commented on src/webhook.lisp:
```diff
@@ -0,0 +1,105 @@
+        (loop for pair in (uiop:split-string header :separator ",")
+              for (key value) = (uiop:split-string pair :separator "=")
+              do (cond ((string= key "t")
+                        (setf timestamp (local-time:unix-to-timestamp (parse-integer value))))
```
Shouldn't `decode-timestamp` be used?

aadcg commented on src/balance-transaction.lisp:
```diff
@@ -22,18 +22,19 @@
+(defmethod initialize-instance :after ((instance balance-transaction) &key data &allow-other-keys)
+  (let ((available-on (gethash :available-on data))
+        (created (gethash :created data))
+        (fee-details (gethash :fee-details data)))
```
It's a huge bummer that we can't use `destructuring-bind` (the reasons being clear).

Nonetheless, I think we can leverage `with-hash-table-iterator`. I don't fancy these endless `gethash` calls. See https://lispcookbook.github.io/cl-cookbook/data-structures.html#traversing-a-hash-table for examples or consult the Hyperspec.

aadcg commented on stripe.asd:
```diff
@@ -1,14 +1,17 @@
+               #:ironclad
                #:local-time
-               #:yason)
+               #:str)
```
Do we really need to pull `str`? It's ok to pull `alexandria` since it's already a dependency of `dexador`.

`str:ends-with-p` can be probably be replaced by `alex:ends-with`. `str:replace-first` can probably be replaced by `substitute`.

boogsbunny replied:
How would you use substitute here?

```
(str:replace-first "item" "-item" object-type)
```

given this case:

```
Special case:
If the object-type is like \"invoiceitem\" (ends with \"item\" but does not
contain an underscore),the function adds a hyphen before \"item\" to match the
naming convention for the corresponding class (e.g., \"invoiceitem\" becomes
\"invoice-item\"). However, if the object-type is like \"subscription_item\"
(contains an underscore before \"item\"), it is left unchanged.
```

boogsbunny replied:
bump @aadcg

aadcg replied:
I'm not sure I understand the question. Doesn't `str:replace-first` fail on the special case you mention above?

aadcg replied:
I had a glimpse at the code. You're making the check on the underscore before calling `str:replace-first`. I don't see why it can't be refactored with routines that don't belong to `str` as to avoid pulling the dependency. I even hinted at the routines you can use to do the refactor up in the thread.

aadcg reviewed (commented):
@boogsbunny I did a more careful review this time and left some ideas.

-------

Note that commit ba28a08 makes use of the JSON library, therefore the webhook commit must come after the migration commit.


-----

I didn't take a close look at the tests but there seems to be some redundancy. For instance, the test below is called 43 times. On the other hand, the test suite didn't catch the issue I mentioned about `decode-timestamp`. Can we do better?

```lisp
(is (local-time:timestamp= (slot-value plan 'stripe::%created)
                           (stripe::decode-timestamp ...)))
```

boogsbunny commented:
Thanks for diving deeper and providing more feedback, I appreciate it.

I'll adjust the PR accordingly soon.

> -----
> 
> 
> 
> I didn't take a close look at the tests but there seems to be some redundancy. For instance, the test below is called 43 times. On the other hand, the test suite didn't catch the issue I mentioned about `decode-timestamp`. Can we do better?
> 
> 
> 
> ```lisp
> 
> (is (local-time:timestamp= (slot-value plan 'stripe::%created)
> 
>                            (stripe::decode-timestamp ...)))
> 
> ```

It's not redundancy. It's intentional to ensure consistency by verifying that each field of an object is parsed correctly for all endpoints.

The issue you mentioned about `decode-timestamp` is a non-issue the way it's used since all timestamps from the Stripe API are unix timestamps measured in seconds (passed as an integer, never as a string).

However, the docstring could benefit from additional context to make this clearer. I can update the docstring or handle the case explicitly to address your concern. I only wanted to point out that the issue you raised doesn't directly apply in this particular case.

boogsbunny commented on src/common.lisp:
```diff
@@ -6,22 +6,69 @@
 
-(defun decode-list (list)
-  (destructuring-bind (&key data has-more &allow-other-keys) list
+(defun decode-list (hash-table)
```
Would it be better if we rename this function since we're decoding a hash table now @aadcg?

boogsbunny commented on tests/balance-transaction-tests.lisp:
```diff
@@ -0,0 +1,79 @@
+  }"
+  "https://docs.stripe.com/api/balance_transactions/list")
+
+(define-mock-query mock-list-balance-history (:type list)
```
Would it better if we rename the argument to `:type` from list to vector for all list endpoints where you receive a vector of objects now @aadcg?

aadcg commented:
@boogsbunny thanks for following along and for you patience. I'm afraid I can only return to this PR in a week. Are you in a hurry? Thanks.

boogsbunny commented:
I'm not in a hurry so take your time!

Let me know if we should rename `decode-list` to something like `decode-table` and change the type for list endpoints from `list` to `vector`.

aadcg commented:
> Let me know if we should rename `decode-list` to something like `decode-table` and change the type for list endpoints from `list` to `vector`.

Since we migrated to a new JSON library, that seems to be reasonable.

I'll reply to your other questions as soon  possible.

boogsbunny commented:
I think I've implemented all the feedback (locally; not pushed yet) minus the commit messages including the new order and handling the case for replacing `item` with `-item` in the renamed `decode-hash-table` function (formerly `decode-list`).

boogsbunny committed:
refactor: migrate JSON library from yason to jzon

This commit migrates the JSON library from yason to jzon for the
following reasons:

1. Easier passing of vector (unsigned byte 8) or stream to jzon for
webhook construction:
- Most people use clack as their web server abstraction layer, thus
it's easier to pass in the payload when constructing webhooks.

2. Improved performance with hash tables instead of plists:
- Destructuring a hash table is generally quicker than a plist for
large JSON objects with nested structures.
- Switching from plists to hash tables allows for quicker access to
elements.

This commit includes the following changes:
- Replace yason usage with jzon for parsing.
- Update the initialize methods for objects to work with hash table
instead of plist including vectors of hash tables when we have a list
of objects.
- Specify the correct return type for a list of objects from
`:type list` to `:type vector`.

boogsbunny committed:
feat: add webhooks

boogsbunny committed:
feat: add test suite

boogsbunny commented:
Hey @aadcg,
I just updated the PR incorporating your feedback.

Please let me know if this looks good!

boogsbunny commented:
bump @aadcg

aadcg commented:
I'm really sorry but I won't be able to review it this week.
